### ISSUE TITLE
Throw errors for illegal states.

### ISSUE THREAD

**Issue by mbostock** [mbostock/queue#41]
- [ ] Calling *queue*.defer after *queue*.await or *queue*.awaitAll.
- [ ] Calling *queue*.await after *queue*.await or *queue*.awaitAll.
- [ ] Calling *queue*.awaitAll after *queue*.await or *queue*.awaitAll.
- [ ] Receiving more than one callback from a task.

**Comment by rclark** [comment]
Hi! Just hit an issue with this where I had been using `.defer` calls after `.awaitAll`.

The use case was one where the list of things to `.defer` was generated asynchronously, so the flow of the application was like

- use `.awaitAll` to provide a function to handle errors
- asynchronously `.defer` several tasks
- once everything has been deferred, `.awaitAll` again to provide a "all finished" function

Could you explain why this should be considered an illegal state?

**Comment by mbostock** [comment]
If you call *queue*.defer after *queue*.await or *queue*.awaitAll, the behavior was nondeterministic and explicitly undefined. The [documentation of *queue*.await and *queue*.awaitAll previously said](https://github.com/mbostock/queue/tree/v1.0.7):

> This method should only be called once, after any tasks have been deferred. If the await callback is set multiple times, or set before a task is deferred, the behavior of the queue is undefined.

If you ignored this advice, it so happened that deferring tasks after registering the await callback behaved similarly to restarting the queue, calling the await callback again when all active deferred tasks completed. But that’s purely a coincidence of the logic and was never intended to be a supported use pattern. Also, it’s inherently a brittle design; the await callback could get called multiple times if the currently-deferred tasks happen to be complete.

This change simply prevents you from relying on the undefined behavior. You must register the await callback after any deferred tasks so that the queue can guarantee that the await callback is invoked exactly once. If you call *queue*.awaitAll before *queue*.defer, then the awaitAll callback is immediately invoked because the queue is empty and by definition complete. If you call *queue*.awaitAll after *queue*.defer, and those tasks have already completed, then the awaitAll callback is likewise immediately invoked; otherwise it waits until the last task completes.

It’s difficult to provide more specific advice based on your description… I’m not sure how you know when “everything has been deferred”, for example. But perhaps you could use one queue to manage when all your tasks have been deferred (registered) and a second queue to manage when all those tasks have been completed.

**Comment by rclark** [comment]
Thanks for the explanation. My implementation that used this behavior was a transform stream that does asynchronous, concurrent work. Very roughly:

```js
var transform = new stream.Transform();
var queue = require('queue-async')(10);

queue.awaitAll(function(err) { transform.emit('error', err); });

transform._transform = function(chunk, enc, callback) {
  queue.defer(doSomething, chunk);
  callback();
});

transform._flush = function(callback) {
  queue.awaitAll(function(err) {
    allDone(err);
    callback();
  });
};
```

So the catch is that I don't know everything has been deferred until someone tells me that it has by calling `.end()` on the stream. 

The implementation might be better off without the initial `.awaitAll` but I would need a way to [avoid this `.concat()` call](https://github.com/mbostock/queue/blob/0b5204991b8973f67a2116ceb7b8843e306a29a8/src/queue.js#L57) which fails when there are tons of deferred tasks.

**Comment by mbostock** [comment]
So you expect some number of asynchronous `transform._transform` calls, followed by a `transform._flush` call, and then the pattern can repeat?

Seems like you’d want to create a new queue after flushing. Maybe like this:

```js
var q = queue(10);

transform._transform = function(chunk, enc, callback) {
  q.defer(doSomething, chunk);
  callback();
});

transform._flush = function(callback) {
  q.awaitAll(function(err) {
    q = queue(10); // Start a new queue.
    allDone(err);
    callback();
  });
};
```

This would also fix the memory leak in the other implementation since a queue always retains its deferred tasks.

**Comment by mbostock** [comment]
Although, that won’t work if you get a transform._transform call while you are flushing. So maybe you want to reset the queue before waiting for the tasks to complete:

```js
transform._flush = function(callback) {
  var q0 = q;
  q = queue(10); // Start a new queue.
  q0.awaitAll(function(err) {
    allDone(err);
    callback();
  });
};
```

**Comment by rclark** [comment]
Thanks! I really appreciate your helping me think this through. To be clear, the transform stream is not reused -- once `_flush()` is called there will be no further `._transform()` calls. 

What I meant about using an initial `.awaitAll` to avoid `.apply()` fail was this:

```
/Users/r/tmp/queue/node_modules/queue-async/build/queue.js:62
      else callback.apply(null, [error].concat(tasks));
                    ^
RangeError: Maximum call stack size exceeded
    at notify (/Users/r/tmp/queue/node_modules/queue-async/build/queue.js:62:21)
    at null.<anonymous> (/Users/r/tmp/queue/node_modules/queue-async/build/queue.js:50:16)
    at timer._onTimeout (timers.js:228:16)
    at Timer.listOnTimeout [as ontimeout] (timers.js:121:15)
```

[Here's the script I used to produce this error](https://gist.github.com/rclark/fbcd5d6255e2251728f7). It crashed around 19900 deferred tasks.

**Comment by mbostock** [comment]
Oh, that’s helpful. It’d be trivial to change the internal `all` flag to true by default and fix this.

**Comment by mbostock** [comment]
Fixed in d8bd9020b38ea97048b845df5765f30428c7e304.
